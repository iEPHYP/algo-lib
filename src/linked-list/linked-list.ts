export interface LinkedListItem<TValue = number> {
  value: TValue;
  nextItem: LinkedListItem<TValue> | null;
  previousItem: LinkedListItem<TValue> | null;
}

export class LinkedList<TValue> {
  public head: LinkedListItem<TValue> | null = null;
  public tail: LinkedListItem<TValue> | null = null;
  private _size = 0;

  private incrementSize() {
    this._size++;
  }

  private decrementSize() {
    if (this._size > 0) {
      this._size--;
    }
  }

  public size() {
    return this._size;
  }

  public empty() {
    return this.size() === 0;
  }

  // O(N)TS
  public toList() {
    const list: TValue[] = [];
    if (this.head) {
      let item = this.head;
      for (let i = 0; i < this.size(); i++) {
        list.push(item.value);
        if (item.nextItem === null) {
          break;
        }
        item = item.nextItem;
      }
    }

    return list;
  }

  // TODO: implement and add test cases
  /** Inserts an element at specified index O(i)T ~ O(N) */
  public insertAt(value: TValue, index: number) {
    throw Error('Not implemented yet');
  }
  /** Inserts an element after specified item O(1)TS */
  public insertAfter(item: LinkedListItem<TValue>, value: TValue) {
    throw Error('Not implemented yet');
  }
  /** Inserts an element before specified item O(1)TS */
  public insertBefore(item: LinkedListItem<TValue>, value: TValue) {
    throw Error('Not implemented yet');
  }
  /** Removes an element at specified index O(i)T ~ O(N) */
  public removeAt(index: number): TValue | null {
    throw Error('Not implemented yet');
  }
  /** Removes an element after specified item O(1)TS */
  public removeAfter(item: LinkedListItem<TValue>): TValue | null {
    throw Error('Not implemented yet');
  }
  /** Removes an element before specified item O(1)TS */
  public removeBefore(item: LinkedListItem<TValue>): TValue | null {
    throw Error('Not implemented yet');
  }

  /** Adds element to the head O(1)TS */
  public unshift(value: TValue) {
    const previousHead = this.head;
    this.head = {
      value,
      nextItem: previousHead,
      previousItem: null,
    };
    if (previousHead) {
      previousHead.previousItem = this.head;
    }
    if (this.size() === 0) {
      this.tail = this.head;
    }
    this.incrementSize();
  }

  /** Removes the first element and returns removed element O(1)TS */
  public shift() {
    const previousHead = this.head;
    this.head = previousHead?.nextItem || null;

    this.decrementSize();

    if (this.size() === 0) {
      this.tail = null;
      this.head = null;
    }

    return previousHead?.value || null;
  }

  /** Adds element to the end O(1)TS */
  public push(value: TValue) {
    if (this.tail) {
      const nextTail = {
        value,
        previousItem: this.tail,
        nextItem: null,
      };
      this.tail.nextItem = nextTail;
      this.tail = nextTail;
    } else {
      this.tail = {
        value,
        previousItem: null,
        nextItem: null,
      };
      this.head = this.tail;
    }
    this.incrementSize();
  }

  /** Removes element from the end and returns removed element O(1)TS */
  public pop(): TValue | null {
    if (this.tail) {
      const tailValue = this.tail.value;
      this.tail = this.tail.previousItem;

      this.decrementSize();

      if (this.size() === 0) {
        this.head = null;
        this.tail = null;
      }

      return tailValue;
    } else {
      return null;
    }
  }
}
