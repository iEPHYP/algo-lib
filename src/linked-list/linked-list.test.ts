import { LinkedList } from './linked-list';

describe('LinkedList', () => {
  let linkedList: LinkedList<number>;

  beforeEach(() => {
    linkedList = new LinkedList<number>();
  });

  it('initializes correctly', () => {
    expect(linkedList.toList()).toMatchSnapshot('initialized Linked List');
  });

  it('adds elements to the beginning with unshift', () => {
    linkedList.unshift(1);
    expect(linkedList.toList()).toMatchSnapshot('unshifts for empty Linked List');
    linkedList.unshift(2);
    expect(linkedList.toList()).toMatchSnapshot('unshifts for Linked List with 1 item');
    linkedList.unshift(3);
    expect(linkedList.toList()).toMatchSnapshot('unshifts for Linked List with 2 items');
  });

  it('removes elements from the beginning with shift', () => {
    expect(linkedList.shift()).toBe(null);
    expect(linkedList.toList()).toMatchSnapshot('shifts for empty Linked List');

    linkedList.unshift(1);
    expect(linkedList.shift()).toBe(1);
    expect(linkedList.toList()).toMatchSnapshot('shifts for Linked List with 1 item');

    linkedList.unshift(1);
    linkedList.unshift(2);
    expect(linkedList.shift()).toBe(2);
    expect(linkedList.toList()).toMatchSnapshot('shifts for Linked List with 2 items');
  });

  it('adds elements to the end with push', () => {
    linkedList.push(1);
    expect(linkedList.toList()).toMatchSnapshot('for empty Linked List');
    linkedList.push(2);
    expect(linkedList.toList()).toMatchSnapshot('for Linked List with 1 item');
    linkedList.push(3);
    expect(linkedList.toList()).toMatchSnapshot('for Linked List with 2 items');
  });

  it('removes elements from the end with pop', () => {
    expect(linkedList.pop()).toBe(null);
    expect(linkedList.toList()).toMatchSnapshot('for empty Linked List');
    linkedList.push(1);
    expect(linkedList.pop()).toBe(1);
    expect(linkedList.toList()).toMatchSnapshot('for Linked List with 1 item');
    linkedList.push(1);
    linkedList.push(2);
    expect(linkedList.pop()).toBe(2);
    expect(linkedList.toList()).toMatchSnapshot('for Linked List with 2 items');
  });

  it('checks if list is empty', () => {
    expect(linkedList.empty()).toBe(true);
    linkedList.push(1);
    expect(linkedList.empty()).toBe(false);
    linkedList.unshift(2);
    expect(linkedList.empty()).toBe(false);
  });

  it('checks the size of the list', () => {
    expect(linkedList.size()).toBe(0);
    linkedList.push(1);
    expect(linkedList.size()).toBe(1);
    linkedList.pop();
    expect(linkedList.size()).toBe(0);
  });

  it('looks as expected after several operations', () => {
    linkedList.unshift(1);
    linkedList.push(2);
    linkedList.push(5);
    linkedList.unshift(10);
    linkedList.pop();
    linkedList.unshift(11);
    linkedList.shift();
    linkedList.shift();
    linkedList.push(22);

    expect(linkedList.toList()).toMatchSnapshot();
  });
});
