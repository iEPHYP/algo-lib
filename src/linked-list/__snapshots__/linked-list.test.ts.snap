// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList adds elements to the beginning with unshift: unshifts for Linked List with 1 item 1`] = `
[
  2,
  1,
]
`;

exports[`LinkedList adds elements to the beginning with unshift: unshifts for Linked List with 2 items 1`] = `
[
  3,
  2,
  1,
]
`;

exports[`LinkedList adds elements to the beginning with unshift: unshifts for empty Linked List 1`] = `
[
  1,
]
`;

exports[`LinkedList adds elements to the end with push: for Linked List with 1 item 1`] = `
[
  1,
  2,
]
`;

exports[`LinkedList adds elements to the end with push: for Linked List with 2 items 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`LinkedList adds elements to the end with push: for empty Linked List 1`] = `
[
  1,
]
`;

exports[`LinkedList initializes correctly: initialized Linked List 1`] = `[]`;

exports[`LinkedList looks as expected after several operations 1`] = `
[
  1,
  2,
  22,
]
`;

exports[`LinkedList removes elements from the beginning with shift: shifts for Linked List with 1 item 1`] = `[]`;

exports[`LinkedList removes elements from the beginning with shift: shifts for Linked List with 2 items 1`] = `
[
  1,
]
`;

exports[`LinkedList removes elements from the beginning with shift: shifts for empty Linked List 1`] = `[]`;

exports[`LinkedList removes elements from the end with pop: for Linked List with 1 item 1`] = `[]`;

exports[`LinkedList removes elements from the end with pop: for Linked List with 2 items 1`] = `
[
  1,
]
`;

exports[`LinkedList removes elements from the end with pop: for empty Linked List 1`] = `[]`;
